KAFKA-6424: QueryableStateIntegrationTest#queryOnRebalance should accept raw text

h314to opened the pull request:
* Remove to .toLowerCase in word count stream
* Add method to read text from file, and move text to resources file
* Minor cleanup: change argument order in assertEquals

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

h314to committed:
KAFKA-6424: QueryableStateIntegrationTest#queryOnRebalance should accept raw text

* Remove to .toLowerCase in word count stream
* Add method to read text from file, and move text to resources file
* Minor cleanup: change argument order in assertEquals

mjsax commented:
Build error is related:
```
Unknown license: /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk9-scala2.12/streams/src/test/resources/data/sampleText.txt
```
We cannot just a file without license header. Please move back to in-memory. Thanks.

h314to commented:
Thanks. I missed that on the output. I'm moving it back to in-memory.

h314to committed:
Remove sampleText.txt and move text back into QueryableStateIntegrationTest

h314to committed:
Add uppercase word to inputValues

mjsax commented:
Call for review @tedyu. I am not sure if this PR addresses the JIRA. As you create the JIRA, you should be able to clarify. Thanks.

tedyu commented:
My intention was to allow the test to read from an external file.
File has bigger capacity which can test related code at higher scale.

mjsax commented:
@tedyu Would the file path/name be hardcoded? Or how to pass in the file name? We need the in-memory fall-back as we cannot just add a file without license header.

An alternative would be, to create the file at startup and read it back during the test.

tedyu commented:
Does data file have to contain license header ?

Hardcoding file path is fine - if the file doesn't exist, we can fall back to the current text.

Thanks

h314to commented:
If we want to have the text in a file and the file must have a license I could just skip the license while reading, or parse the license lines, adding them to the word count stream, and adjust the test to accommodate this change.

tedyu commented:
Either way is fine.
Probably skipping the header is less work for you.

Thanks

h314to committed:
Add method to read file again and move text to resources

tedyu commented on streams/src/test/resources/data/sampleText.txt:
```diff
@@ -0,0 +1,28 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
```
Can you add a comment here saying not to add more line(s) above this line ?

mjsax replied:
Thinking about this twice, it might be preferable to not add this file because the parsing logic skipping the header is non-intuitive. We could instead keep the in-memory data for regular test, but add a check if this file exist and if yes, use the file instead of in-memory data. WDYT?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -167,20 +185,7 @@ public int compare(final KeyValue<String, Long> o1,
-            "eat sleep rave repeat",
-            "one jolly sailor",
-            "king of the world");
+        inputValues = readInputValues("data/sampleText.txt", 15);
```
Thinking about this, it might be better to keep in-memory as default for the case the file does not exist and read the file if it exist overwriting the in-memory data. This keep the test itself cleaner and avoid the header skipping logic that is quite unintuitive.

For large files, I am wondering, if reading the whole file into an array is good enough... What file sized did you have in mind @tedyu ?

tedyu replied:
About 5x to 10x the size of what the array holds.

h314to committed:
Delete input file from resources and default to original test list.

h314to commented:
I've done a few things to try to address your comments. I removed the header skipping logic and now default to a list of strings that is enough to run the tests.

I documented the method to make what it is doing clear, and changed the default file name to `resources/QueryableStateIntegrationTest/inputValues.txt` which I think is a bit more obvious.

mjsax reviewed (approved):
LGTM.

\cc @tedyu Call for review.

tedyu commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -133,6 +135,39 @@ private void createTopics() throws InterruptedException {
+                input.add(line);
+            }
+        } catch (Exception e) {
+            input = Arrays.asList(
```
Add a log saying that internal strings are used since inputValues.txt is absent.

h314to committed:
Log that the default inputValues list is being used

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -133,6 +139,40 @@ private void createTopics() throws InterruptedException {
+     */
+    private List<String> getInputValues() {
+        List<String> input = new ArrayList<>();
+        ClassLoader classLoader = getClass().getClassLoader();
```
Just some more nits. Can you add `final` wherever possible: `ClassLoader`, `String filename`, `BufferedReader`, `for(final String...)`, `Exception`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -133,6 +139,40 @@ private void createTopics() throws InterruptedException {
+    private List<String> getInputValues() {
+        List<String> input = new ArrayList<>();
+        ClassLoader classLoader = getClass().getClassLoader();
+        String fileName = "QueryableStateIntegrationTest/inputValues.txt";
```
Use `File.separator` instead of `/`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java:
```diff
@@ -133,6 +139,40 @@ private void createTopics() throws InterruptedException {
+                input.add(line);
+            }
+        } catch (Exception e) {
+            log.warn("Unable to read '{}'. Using default inputValues list", "resources/" + fileName);
```
as above: avoid `/`

Update to
```
log.warn("Unable to read '{}{}{}'. Using default inputValues list", "resources", File.seperator, fileName);
```

h314to replied:
Yes, that's certainly safer. Thanks. I also changed it in fileName and added `final` as requested.

h314to committed:
Add final and use File.separator instead of /

mjsax reviewed (approved):
Thanks for the update @h314to . Will merge after we get a green build.

mjsax commented:
Retest this please.

mjsax commented:
Retest this please.

mjsax commented:
Merged to `trunk`

Thanks for the PR @h314to !

h314to commented:
Great! Thanks for the helpful reviews.
